lib/experimental/graph.nim

```diff
@@ -0,0 +1,428 @@
+    let top = stack[^1].pop
+
+    if top notin yielded:
+      yield (top, stack.len())
+      yielded.incl top
+
+    var buf: seq[Id]
+    for (src, node) in graph.outEdges(top):
+      if node notin visited:
+        visited.incl node
+        buf.add node
+
+    if stack[^1].len() == 0:
+      discard stack.pop()
+
+    if 0 < len(buf):
+      stack.add buf
+
+iterator breadthFirst*[Id](graph: HGraph[Id], root: Id): Id =
+
+  ## Perform breadth-first iteration of parent graph, starting from `node`
+  var que = initDeque[Id]()
+  que.addLast(root)
+  var visited: HashSet[Id]
+  while 0 < len(que):
+    let top = que.popFirst()
+    yield top
+    for (src, target) in graph.outEdges(top):
+      if target notin visited:
+        visited.incl target
+        que.addLast(target)
+
+
+proc getNoIncoming*[Id](graph: HGraph[Id]): seq[Id] =
+  ## Get all nodes with zero in degree of zero
+  let degrees = computeInDegrees(graph)
+  for node in graph.nodes:
+    if degrees[node] == 0:
+      result.add node
+
+proc topologicalOrdering*[Id](
+    graph: HGraph[Id], start: seq[Id] = @[]): seq[Id] =
+  ## Return graph nodes in topological ordering if possible. Otherwise (if
+  ## graph contains cycles raise `HGraphCyclesError`)
+  var active = toHashSet(start)
+  if active.len() == 0:
+    active = toHashSet(getNoIncoming(graph))
+
+  var graph = graph
+
+  while 0 < len(active):
+    let node = active.pop
+    result.add node
+    for (src, target) in graph.outEdges(node):
+      graph.removeEdge((src, target))
+      if graph.inDeg(target) == 0:
+        active.incl target
+
+  if 0 < graph.edgeCount:
+    raise HGraphCyclesError(
+      msg: "Topological sorting is impossible, graph contains cycles")
+
+
+template findMaxIt*(s: untyped, op: untyped): untyped =
+  var res: int = 0
+  var idx = 0
+  var prev: typeof((let it {.inject.} = s[0]; op))
+
+  for it {.inject.} in s:
+    let val = op
+    if val > prev:
+      res = idx
+
+    inc idx
+
+  res
+
+template findIt*(s: typed, op: untyped): int =
+  ## Find first element of the sequence for which `op` evaluates as true
+  ## and return it's index. If no such element is found return `-1`
+
+  var result = -1
+  for idx, it {.inject.} in s:
+    if op: result = idx; break
+
+  result
+
+
+proc graphEasyRepr*[Id](
+    graph: HGraph[Id], name: proc(id: Id): string): string =
+
+  for (src, dst) in edgePairs(graph):
+    result.add "[ $# ] $# [ $# ]\n" % [
+      name(src),
+      "->",
+      name(dst)
+    ]
+
+proc findCycles*[Id](
+    graph: HGraph[Id],
+    ignoreSelf: bool = false,
+    overrideDirected: bool = false,
+    yieldMultiedge: bool = false
+  ): seq[seq[Id]] =
+
+  var
+    visited: HashSet[Id]
+    stack: seq[Id]
+    resCycles: seq[seq[Id]]
+
+  proc printCycle(vertex: Id) =
+    var stack2: seq[Id]
+    stack2.add(stack.pop())
+
+    while stack2[^1] != vertex:
+      stack2.add(stack.pop())
+
+    var path: seq[Id]
+    while 0 < len(stack2):
```

Comment:
I get the idea behind putting the constant item first when it's an assignment to avoid errors in other languages. I find I have to continually flip it around in my head and then see if it makes sense. It's just not how most folks are conditioned to think.